File: Easy/234-PalindromeLinkedList.cpp
#include <iostream>
#include <vector>
using namespace std;

// Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
private:
    bool reverseCheck(vector<int>& toCheck, ListNode* checkPtr) {
        int i;
        ListNode* currPtr = checkPtr;
        // first check if it's an even-numbered palindrome
        for (i = toCheck.size() - 1; i >= 0 && currPtr != nullptr; --i) {
            if (toCheck[i] != currPtr->val) {
                break;
            }
            currPtr = currPtr->next;
        }
        if (i == -1 && currPtr == nullptr) {
            // then it's a valid even-numbered palindrome
            return true;
        }

        // now check if it's an odd-numbered palindrome
        currPtr = checkPtr;
        for (i = toCheck.size() - 2; i >= 0 && currPtr != nullptr; --i) {
            if (toCheck[i] != currPtr->val) {
                break;
            }
            currPtr = currPtr->next;
        }

        return (i == -1 && currPtr == nullptr);
    }
public:
    bool isPalindrome(ListNode* head) {
        // check the trivial case first
        if (head->next == nullptr) {
            return true;
        }
        vector<int> vals;
        ListNode* currPtr = head;
        while (currPtr != nullptr) {
            if (reverseCheck(vals, currPtr)) {
                return true;
            }
            vals.push_back(currPtr->val);
            currPtr = currPtr->next;
        }
        return false;
    }
};

File: Easy/234-PalindromeLinkedList.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        # first use the tortoise and the hare algorithm to get
        # a pointer to the midpoint of the linked list
        slow = head
        fast = head
        
        while fast != None and fast.next != None:
            # move slow along once and fast along twice
            slow = slow.next
            fast = fast.next.next
        
        # save the midpoint
        midpoint = slow
        
        # now reverse the second half of the linked list
        prev = None
        while slow != None:
            
